filestream/filestream.go

```diff
@@ -39,14 +42,34 @@ func New(ctx context.Context, file_name string) (*FileStream, error) {
 	}
 
 	return &FileStream{
-		file_name: file_name,
-		reader:    file,
-		watcher:   watcher,
-		ctx:       ctx,
-		done:      make(chan bool),
+		file_name:   file_name,
+		tailEnabled: tail,
+		reader:      file,
+		watcher:     watcher,
+		ctx:         ctx,
+		done:        make(chan bool),
+		logComplete: make(chan bool),
 	}, nil
 }
 
+//-------------------------------------------------------------------------------------------------
+func (s *FileStream) TailEnabled() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.tailEnabled
+}
+
+func (s *FileStream) DisableTail() {
+	if !s.TailEnabled() {
+		return
+	}
+
+	s.mu.Lock()
```

Comment:
This is another case of you acquiring the lock multiple times and actually not getting an atomic operation out of it.

In this case, tail can be enabled on your first check, and then disabled by a concurrent call to `DisableTail()` before you re-acquire the mutex, resulting in a double close of `s.logComplete`, which will panic.

Reply:
Great catch, fixed that by switching to direct reads of the flag under a lock